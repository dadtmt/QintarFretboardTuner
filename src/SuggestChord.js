import React, { useState } from 'react'
import { entries } from '@tonaljs/chord-dictionary'
import Autosuggest from 'react-autosuggest'

export const chordSuggestions = entries()
  .map(chordType => chordType.name)
  .filter(chordType => chordType)

export const getSuggestions = value => {
  const inputValue = value.trim().toLowerCase()
  return chordSuggestions
    .map(suggestion => ({
      name: suggestion,
      valueIndex: suggestion.indexOf(inputValue)
    }))
    .filter(({ valueIndex }) => valueIndex >= 0)
    .sort((a, b) => a.valueIndex - b.valueIndex)
    .map(({ name }) => name)
}

const getSuggestionValue = suggestion => suggestion

const renderSuggestion = suggestion => <div>{suggestion}</div>

function SuggestChord() {
  const [value, setValue] = useState('')
  const [suggestions, setSuggestions] = useState(chordSuggestions)
  const onChange = (event, { newValue }) => setValue(newValue)
  const onSuggestionsFetchRequested = ({ value }) =>
    setSuggestions(getSuggestions(value))
  const onSuggestionsClearRequested = () => setSuggestions(chordSuggestions)
  const inputProps = {
    placeholder: 'Choose a chord',
    value,
    onChange
  }

  return (
    <Autosuggest
      suggestions={suggestions}
      onSuggestionsFetchRequested={onSuggestionsFetchRequested}
      onSuggestionsClearRequested={onSuggestionsClearRequested}
      getSuggestionValue={getSuggestionValue}
      renderSuggestion={renderSuggestion}
      inputProps={inputProps}
    />
  )
}

export default SuggestChord
