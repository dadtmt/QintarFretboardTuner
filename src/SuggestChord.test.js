import React from 'react'
import ReactDOM from 'react-dom'
import renderer, { act } from 'react-test-renderer'
import { toMatchDiffSnapshot } from 'snapshot-diff'
import SuggestChord, { chordSuggestions, getSuggestions } from './SuggestChord'

expect.extend({ toMatchDiffSnapshot })

describe('chordSuggestions', () => {
  it('suggest all chord names', () => {
    expect(chordSuggestions).toMatchSnapshot()
  })
})

describe('getSuggestions', () => {
  it('get suggestions for value m', () => {
    expect(getSuggestions('m')).toMatchSnapshot()
  })
  it('get suggestions for empty value', () => {
    expect(getSuggestions('')).toMatchSnapshot()
  })
  it('get suggestions for value wav', () => {
    expect(getSuggestions('wav')).toMatchSnapshot()
  })
})

describe('SuggestChord', () => {
  it('renders correctly', () => {
    const component = renderer.create(<SuggestChord />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders without crashing', () => {
    const div = document.createElement('div')
    ReactDOM.render(<SuggestChord />, div)
    ReactDOM.unmountComponentAtNode(div)
  })
})
