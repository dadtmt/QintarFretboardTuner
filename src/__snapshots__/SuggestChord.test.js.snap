// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SuggestChord renders correctly 1`] = `
<div
  aria-expanded={false}
  aria-haspopup="listbox"
  aria-owns="react-autowhatever-1"
  className="react-autosuggest__container"
  role="combobox"
>
  <input
    aria-activedescendant={null}
    aria-autocomplete="list"
    aria-controls="react-autowhatever-1"
    autoComplete="off"
    className="react-autosuggest__input"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    placeholder="Choose a chord"
    type="text"
    value=""
  />
  <div
    className="react-autosuggest__suggestions-container"
    id="react-autowhatever-1"
    role="listbox"
  />
</div>
`;

exports[`chordSuggestions suggest all chord names 1`] = `
Array [
  "fifth",
  "suspended 4th",
  "suspended 4th seventh",
  "augmented",
  "major seventh b6",
  "augmented seventh",
  "major",
  "major seventh",
  "dominant seventh",
  "sixth",
  "lydian",
  "lydian dominant seventh",
  "minor",
  "minor/major seventh",
  "minor seventh",
  "minor sixth",
  "diminished",
  "half-diminished",
  "diminished seventh",
  "dominant #9",
  "suspended 2nd",
  "eleventh",
  "major ninth",
  "dominant ninth",
  "sixth/ninth",
  "major thirteenth",
  "dominant thirteenth",
  "major #11 (lydian)",
  "minor ninth",
  "minor thirteenth",
  "minor eleventh",
  "suspended 4th b9",
  "altered",
  "dominant b9",
]
`;

exports[`getSuggestions get suggestions for empty value 1`] = `
Array [
  "fifth",
  "suspended 4th",
  "suspended 4th seventh",
  "augmented",
  "major seventh b6",
  "augmented seventh",
  "major",
  "major seventh",
  "dominant seventh",
  "sixth",
  "lydian",
  "lydian dominant seventh",
  "minor",
  "minor/major seventh",
  "minor seventh",
  "minor sixth",
  "diminished",
  "half-diminished",
  "diminished seventh",
  "dominant #9",
  "suspended 2nd",
  "eleventh",
  "major ninth",
  "dominant ninth",
  "sixth/ninth",
  "major thirteenth",
  "dominant thirteenth",
  "major #11 (lydian)",
  "minor ninth",
  "minor thirteenth",
  "minor eleventh",
  "suspended 4th b9",
  "altered",
  "dominant b9",
]
`;

exports[`getSuggestions get suggestions for value m 1`] = `
Array [
  "major seventh b6",
  "major",
  "major seventh",
  "minor",
  "minor/major seventh",
  "minor seventh",
  "minor sixth",
  "major ninth",
  "major thirteenth",
  "major #11 (lydian)",
  "minor ninth",
  "minor thirteenth",
  "minor eleventh",
  "dominant seventh",
  "diminished",
  "diminished seventh",
  "dominant #9",
  "dominant ninth",
  "dominant thirteenth",
  "dominant b9",
  "augmented",
  "augmented seventh",
  "half-diminished",
  "lydian dominant seventh",
]
`;

exports[`getSuggestions get suggestions for value wav 1`] = `Array []`;
